### PR TITLE
Add support for convtranspose 2d

### PR DESCRIPTION
**Pull request by zheng-ningxin**
Support the ConvTranspose2d for speedup in this PR. The pruning of convtranspose2d will be supported in the future.

### CODE REVIEW

**Comment by NiuChan1301** [comment]
>Add support for ConvTranspose2d.

File: nni/compression/pytorch/speedup/compress_modules.py [hunk 1]
```diff
@@ -152,24 +163,120 @@ def replace_conv2d(conv, mask):
...
             end = (groupid + 1) * input_step
-            current_input_index = list(filter(lambda x: start <= x and x < end, in_channels_index.tolist()))
+            current_input_index = list(
+                filter(lambda x: start <= x and x < end, in_channels_index.tolist()))
             if not current_input_index:
                 # there is no kept channel in current group
-                continue
+                # TODO bug here, the groups is directly get from conv.groups, if the whole group is removed,
+                # then the number of groups in the new_conv also need to change
+                raise Exception(
```
Comment (QuanluZhang) [inline nni/compression/pytorch/speedup/compress_modules.py:172]: do we support depthwise conv?

File: nni/compression/pytorch/speedup/infer_shape.py [hunk 2]
```diff
@@ -228,6 +230,7 @@ def __repr__(self):
 infer_from_mask = {
     'BatchNorm2d': lambda module_masks, mask: batchnorm2d_mask(module_masks, mask),
     'Conv2d': lambda module_masks, mask: conv2d_mask(module_masks, mask),
+    'Conv2dTranspose2d': lambda module_masks, mask: convtranspose2d_mask(module_masks, mask),
```
Comment (QuanluZhang) [inline nni/compression/pytorch/speedup/infer_shape.py:233]: Conv2dTranspose2d?

File: nni/compression/pytorch/speedup/compress_modules.py (hunk 1 above)
Comment (zheng-ningxin) [reply nni/compression/pytorch/speedup/compress_modules.py:172]: Yes, we do. We can only support pruning the whole group of filters for the depthwise conv.

File: nni/compression/pytorch/speedup/compress_modules.py [hunk 3]
```diff
@@ -152,24 +163,120 @@ def replace_conv2d(conv, mask):
...
+    new_convtrans.to(convtrans.weight.device)
+    tmp_weight_data = None
+    if mask.input_mask is not None:
+        # in convtranspose2d we need to select the input channel first
+        tmp_weight_data = torch.index_select(
+            convtrans.weight.data, 0, in_channels_index)
+    else:
+        tmp_weight_data = convtrans.weight.data
+    # we need to handle the output channel group by group like the conv layer
+    out_step = int(convtrans.out_channels) / convtrans.groups
```
Comment (J-shang) [inline nni/compression/pytorch/speedup/compress_modules.py:251]: the brackets might be wrong?

File: nni/compression/pytorch/speedup/compress_modules.py [hunk 4]
```diff
@@ -152,24 +163,120 @@ def replace_conv2d(conv, mask):
...
+
+def replace_convtranspose2d(convtrans, mask):
+    """
+    We need a anthor replace function for
+    convtranspose2d, because the layout of
+    the weight is different from traditional
+    conv layers. The layout of the weight is [N_in, N_out, ksize_1, ksize_2]
+    Parameters
+    ----------
+    conv : torch.nn.ConvTranspose2d
```
Comment (J-shang) [inline nni/compression/pytorch/speedup/compress_modules.py:203]: `convtrans: torch.nn.ConvTranspose2d`?

File: nni/compression/pytorch/speedup/infer_shape.py [hunk 5]
```diff
@@ -988,3 +1016,74 @@ def conv2d_outshape(module_masks, mask):
...
+    assert mask.mask_index[2] is None
+    assert mask.mask_index[3] is None
+
+    if module_masks.output_mask is None:
+        module_masks.output_mask = mask
+    else:
+        # mask conflict should be solved by fix_mask_conflict before speedup
+        # mask and module_masks.output_mask may have different number of dimensions
+        # since they could be passed by linear or conv2d
+        assert all(
```
Comment (J-shang) [inline nni/compression/pytorch/speedup/infer_shape.py:1072]: why we use `all()` here?

File: nni/compression/pytorch/speedup/infer_shape.py (hunk 2 above)
Comment (zheng-ningxin) [reply nni/compression/pytorch/speedup/infer_shape.py]: A typo,  fixed thanks.

File: nni/compression/pytorch/speedup/compress_modules.py (hunk 4 above)
Comment (zheng-ningxin) [reply nni/compression/pytorch/speedup/compress_modules.py]: Fixed, thanks.

File: nni/compression/pytorch/speedup/compress_modules.py (hunk 3 above)
Comment (zheng-ningxin) [reply nni/compression/pytorch/speedup/compress_modules.py]: fixed,  thanks.

File: nni/compression/pytorch/speedup/compress_modules.py [hunk 6]
```diff
@@ -152,24 +163,120 @@ def replace_conv2d(conv, mask):
...
+    new_convtrans.to(convtrans.weight.device)
+    tmp_weight_data = None
+    if mask.input_mask is not None:
+        # in convtranspose2d we need to select the input channel first
+        tmp_weight_data = torch.index_select(
+            convtrans.weight.data, 0, in_channels_index)
+    else:
+        tmp_weight_data = convtrans.weight.data
+    # we need to handle the output channel group by group like the conv layer
+    out_step = int(convtrans.out_channels) / int(convtrans.groups)
```
Comment (J-shang) [inline nni/compression/pytorch/speedup/compress_modules.py:251]: `out_step` is a float, maybe `out_step = int(convtrans.out_channels) // int(convtrans.groups)` ?

Comment (zheng-ningxin) [reply nni/compression/pytorch/speedup/compress_modules.py]: Updated, pls review again.

**Review by J-shang — APPROVED** [review]

File: nni/compression/pytorch/speedup/compress_modules.py [hunk 7]
```diff
@@ -152,24 +163,120 @@ def replace_conv2d(conv, mask):
...
-        new_conv.bias.data.copy_(conv.bias.data if tmp_bias_data is None else tmp_bias_data)
+        new_conv.bias.data.copy_(
+            conv.bias.data if tmp_bias_data is None else tmp_bias_data)
 
     return new_conv
+
+
+def replace_convtranspose2d(convtrans, mask):
+    """
+    We need a anthor replace function for
```
Comment (linbinskn) [inline nni/compression/pytorch/speedup/compress_modules.py:197]: need another

**Review by chicm-ms — APPROVED** [review]

File: nni/compression/pytorch/speedup/compress_modules.py (hunk 7 above)
Comment (zheng-ningxin) [reply nni/compression/pytorch/speedup/compress_modules.py]: fixed, thanks.

File: nni/compression/pytorch/speedup/compress_modules.py [hunk 8]
```diff
@@ -152,24 +163,120 @@ def replace_conv2d(conv, mask):
...
 
     if conv.bias is not None:
-        new_conv.bias.data.copy_(conv.bias.data if tmp_bias_data is None else tmp_bias_data)
+        new_conv.bias.data.copy_(
+            conv.bias.data if tmp_bias_data is None else tmp_bias_data)
 
     return new_conv
+
+
+def replace_convtranspose2d(convtrans, mask):
```
Comment (linbinskn) [inline nni/compression/pytorch/speedup/compress_modules.py:195]: When mask is None, may be just return convtrans will be simpler?

**Review by QuanluZhang — APPROVED** [review]

File: nni/compression/pytorch/speedup/compress_modules.py (hunk 8 above)
Comment (zheng-ningxin) [reply nni/compression/pytorch/speedup/compress_modules.py:195]: you mean add the following code snippet at the start?
```
if mask is None:
   return convtrans
```

Comment (linbinskn) [reply nni/compression/pytorch/speedup/compress_modules.py:195]: Yeah. If both of mask.input_mask and mask.output_mask are None, we will still create a new module and copy related parameters but it is unnecessary. But I am not sure whether this situation exists or not.

Comment (zheng-ningxin) [reply nni/compression/pytorch/speedup/compress_modules.py:195]: https://github.com/microsoft/nni/blob/622e3331c946defe3be529ee7584d3f713f4c70f/nni/compression/pytorch/speedup/compressor.py#L151 
We only replace the modules that have inferred masks.
